packages/workbox-background-sync/lib/Queue.mjs

```diff
@@ -0,0 +1,297 @@
+    let allReplaysSuccessful = true;
+
+    for (const [key, storableRequest] of storableRequestsInQueue) {
+      this._runCallback('requestWillReplay', storableRequest);
+
+      const request = storableRequest.toRequest();
+      let response;
+
+      try {
+        response = await fetch(request);
+      } catch (error) {
+        this._runCallback('replayDidFail', {error, request});
+        allReplaysSuccessful = false;
+        continue;
+      }
+
+      this._runCallback('requestDidReplay', {request, response});
+      successfullyReplayedRequests.push({request, response});
+
+      // Remove the request from IndexedDB asynchronously (don't await).
+      // TODO(philipwalton): in the unlikely event that the delete fails,
+      // this request may be replayed again. Do we want to handle this case?
+      this._removeRequest(key);
+    }
+
+    if (allReplaysSuccessful) {
+      this._runCallback('allRequestsDidReplay', successfullyReplayedRequests);
+    } else {
+      this._registerSync();
+    }
+  }
+
+  /**
+   * Gets all requests in the object store matching this queue's name.
+   *
+   * @private
+   * @return {Promise<Array>}
+   */
+  async _getStorableRequestsInQueue() {
+    const db = await this._getDb();
+    const storableRequests = [];
+
+    for (const [key, entry] of (await db.getAll()).entries()) {
+      if (entry.queueName == this._name) {
+        // Requests older than `maxRetentionTime` should be ignored.
+        const storableRequest = new StorableRequest(entry.storableRequest);
+
+        if (Date.now() - storableRequest.timestamp > this._maxRetentionTime) {
+          // No need to await this since it can happen in parallel.
+          this._removeRequest(key);
+          continue;
+        }
+
+        storableRequests.push([key, storableRequest]);
+      }
+    }
+
+    return storableRequests;
+  }
+
+  /**
+   * Gets a reference to the IndexedDB object store for queued requests.
+   *
+   * @private
+   * @return {Promise<DBWrapper>}
+   */
+  _getDb() {
+    return indexedDBHelper.getDB(
+        DB_NAME, OBJECT_STORE_NAME, {autoIncrement: true});
+  }
+
+  /**
+   * Removes a request from IndexedDB for the specified key.
+   *
+   * @private
+   * @param {string} key
+   */
+  async _removeRequest(key) {
+    const db = await this._getDb();
+    await db.delete(key);
+  }
+
+  /**
+   * Runs the passed callback if it exists.
+   *
+   * @private
+   * @param {string} name The name of the callback on this._callbacks.
+   * @param {...*} args The arguments to invoke the callback with.
+   */
+  _runCallback(name, ...args) {
+    if (typeof this._callbacks[name] == 'function') {
+      this._callbacks[name].apply(null, args);
+    }
+  }
+
+  /**
+   * In sync-supporting browsers, this adds a listener for the sync event.
+   * In non-sync-supporting browsers, this will retry the queue on service
+   * worker startup.
+   *
+   * @private
+   */
+  _addSyncListener() {
```

Comment:
I'm not sure I understand. Can you give more context?

Looking at those examples, they both expose a single instance of a class, whereas background sync exposes the class itself, and it needs a unique listener per instance.

Reply:
https://github.com/GoogleChrome/workbox/blob/v3/packages/workbox-routing/default-export.mjs#L22 is an example of what I did for the `Router` class.

I had trouble wrapping my head around it as well, but once you get a feel for how the `default` exports are likely be used in the context of `workbox-sw`, it makes more sense.

Reply:
I still don't think this applies to the background sync case since I don't instantiate anything in the default export.

Reply:
That should be okay by me, let's see if @gauntface has any concerns. He's given this setup a lot more thought that I have.

Reply:
I still have concerns but this module is also too big to cover off everything. I'll play around with it and see if it makes to refactor after.

Reply:
Ok, I'd be happy to look at a suggestion for how it could be refactored, but since I don't understand what you're asking me to do, I can't really do it myself.